src/Mvc/test/WebSites/BasicWebSite/Controllers/BindPropertiesWithValidationController.cs

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.ComponentModel.DataAnnotations;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace BasicWebSite
+{
+    [BindProperties(SupportsGet = true)]
+    public class BindPropertiesWithValidationController : Controller
+    {
+        [Required(ErrorMessage = "User name is required.")]
+        public string UserName { get; set; }
+
+        [Required]
+        public string Password { get; set; }
+
+        [Compare(nameof(Password), ErrorMessage = "Password and confirm password do not match.")]
```

Comment:
Is this defence in depth i.e. doesn't `[Compare]` already work fine in controllers?

Reply:
It doesn't if the bound properties are on the controller. It's rare that users do this though which is likely why we haven't heard much about this.